src/flatten.rs

```diff
@@ -0,0 +1,623 @@
+    /// Target file to flatten.
+    target: PathBuf,
+    /// Sources including only target and it dependencies (imports of any depth).
+    sources: Sources,
+    /// Vec of (path, ast) pairs.
+    asts: Vec<(PathBuf, SourceUnit)>,
+    /// Sources in the order they should be written to the output file.
+    ordered_sources: Vec<PathBuf>,
+}
+
+impl Flattener {
+    /// Compilation output is expected to contain all artifacts for all sources.
+    /// Flattener caller is expected to resolve all imports of target file, compile them and pass
+    /// into this function.
+    pub fn new(project: &Project, output: &ProjectCompileOutput, target: &Path) -> Result<Self> {
+        let input_files = output
+            .artifacts_with_files()
+            .map(|(file, _, _)| PathBuf::from(file))
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        let sources = Source::read_all_files(input_files)?;
+        let graph = Graph::resolve_sources(&project.paths, sources)?;
+
+        let ordered_deps = collect_ordered_deps(&target.to_path_buf(), &project.paths, &graph)?;
+
+        let sources = Source::read_all(&ordered_deps)?;
+
+        // Convert all ASTs from artifacts to strongly typed ASTs
+        let mut asts: Vec<(PathBuf, SourceUnit)> = Vec::new();
+        for (path, ast) in output.artifacts_with_files().filter_map(|(path, _, artifact)| {
+            if let Some(ast) = artifact.ast.as_ref() {
+                if sources.contains_key(&PathBuf::from(path)) {
+                    Some((path, ast))
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }) {
+            asts.push((PathBuf::from(path), serde_json::from_str(&serde_json::to_string(ast)?)?));
+        }
+
+        Ok(Flattener { target: target.into(), sources, asts, ordered_sources: ordered_deps })
+    }
+
+    /// Flattens target file and returns the result as a string
+    ///
+    /// Flattening process includes following steps:
+    /// 1. Find all file-level definitions and rename references to them via aliased or qualified
+    ///    imports.
+    /// 2. Find all duplicates among file-level definitions and rename them to avoid conflicts.
+    /// 3. Remove all imports.
+    /// 4. Remove all pragmas except for the ones in the target file.
+    /// 5. Remove all license identifiers except for the one in the target file.
+    pub fn flatten(&self) -> String {
+        let mut updates = Updates::new();
+
+        let top_level_names = self.rename_top_level_definitions(&mut updates);
+        self.rename_contract_level_types_references(&top_level_names, &mut updates);
+        self.remove_imports(&mut updates);
+        let target_pragmas = self.process_pragmas(&mut updates);
+        let target_license = self.process_licenses(&mut updates);
+
+        let result = FlatteningResult::new(self, updates, target_pragmas, target_license);
```

Comment:
FlattenedResult only borrows from `Flattener`, right?

let's move this to `fn flatten_result` so that this function becomes `self.flatten_result().get_flattened_target`
if possible